homeassistant/components/neptune_apex/light.py

```diff
@@ -0,0 +1,131 @@
+"""Support outlet status/control from Neptune Apex."""
+import logging
+
+import pynepsys
+
+from homeassistant.components.light import ATTR_EFFECT, SUPPORT_EFFECT, Light
+
+from . import NEPTUNE_APEX, NEPTUNE_APEX_COORDINATOR
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Parse through the Apex outlets and creates Outlet entities."""
+    apex = hass.data[NEPTUNE_APEX]
+    coordinator = hass.data[NEPTUNE_APEX_COORDINATOR]
+
+    async_add_entities(
+        Outlet(coordinator, apex, name)
+        for i, name in enumerate(coordinator.data.outlets)
+    )
+
+
+class Outlet(Light):
+    """Shoehorn the Apex outlet concept into a hass Light entity.
+
+    This feels gross, but the more you play with Apex, you realize that once you get past
+    a certain point of revulsion, there is beauty in it. And madness.
+    """
+
+    def __init__(self, coordinator, apex: pynepsys.Apex, name):
+        """Initialize this outlet, storing reference to parent Apex."""
+        self.coordinator = coordinator
+        self._name = name
+        self.apex = apex
+        self.outlet: pynepsys.Outlet = apex.outlets[name]
+
+    @property
+    def name(self):
+        """Name of this outlet (From the Apex), prefixed."""
+        return "apex." + self._name
+
+    @property
+    def unique_id(self):
+        """Name of this outlet as defined in the Apex."""
+        return self.name
+
+    @property
+    def is_on(self):
+        """Return True if the outlet is manually forced on or in AUTO mode and on."""
+        return self.outlet.is_on()
+
+    @property
+    def supported_features(self):
+        """We support effects as a hack to allow AUTO mode to exist."""
+        return SUPPORT_EFFECT
+
+    @property
+    def effect(self):
+        """Return the current state of the outlet.
+
+        ON, OFF, AUTO are the useful effects for this integration, but you can
+        read profile names for pumps and lights via this property as well.
+        """
+        return self.outlet.state
+
+    @property
+    def effect_list(self):
+        """Return the profiles you can actually put this outlet into.
+
+        AUTO represents automatic mode, ON/OFF being manually forced on/off
+        ignoring outlet programming.
+        """
+        return ["AUTO", "ON", "OFF"]
```

Comment:
These don't seem like effects? Why are we using the light entity if this is not a light device?

Reply:
I'm not sure there's a better analogue to the HA light for the Apex Outlet. Not to go too historical, but the older-generation Aquacontrollers used X10 to control lights and pumps, so that's where the concept of 'outlet' came from. The Apex has the obvious ON/OFF states, which are manual modes, but then has a third mode, 'AUTO'. This then leads to two more modes, AUTO-ON (AON) and AUTO-OFF (AOF). Pushing an outlet into AUTO and then waiting for the controller to update to AON/AOF is the only path. This works really well for regular things controlled via their EnergyBars, which have 4 or 8 physical plugs each. Cool, everything seems good for binary control. Except there are other outputs that are still called outlets, but are variable, some of 0-10V, some are 0-100%. And you cannot actually tell an outlet to go to 50% or 5V or whatever. You must enable a profile. Profiles specify start and end states, ramp times, cycle types, etc.

So long story short, when an outlet is binary, it has five states, but 2 are ON, 2 are OFF, and 1 is unknown. If it's analogue, it has two states, OFF and NOT-OFF, as you can't actually tell what % or voltage it's currently at. HASS light seemed like the only type that could encompass all this and actually work. You can manually turn things on/off, and you can put the outlet back into AUTO mode by setting the effect to AUTO. 

Hopefully my logic checks?

Reply:
Can we instead make this a switch and add a custom service to activate and disable the auto mode? It sounds like the state of the device is still binary, except when we don't know state. So the switch should work for reporting state.

Reply:
Oh, one more fun thing... Putting an outlet into AUTO is also the way to get it to set a profile... It's via its own internal program. I forgot you can't force them straight to whatever. Profiles are user-defined names.

Reply:
I'm not sure that'd work with what I know of switch... As an example, right now I have an alert that fires if I leave a particular light on in manual mode (so 'ON') for longer than 30 minutes. But that light will correctly turn itself on for 8 hours a day and not trigger the alert (AON). It also allows you to see if you're in auto or manual mode via the regular interface, and also to see what state a pump is in (for instance right now my Vortech pump is in 'TBL' mode)

Reply:
We could add a state attribute with the auto mode state?

Reply:
Is there a good example component that I can copy?

Reply:
Here's the docs for adding a custom entity service:
https://developers.home-assistant.io/docs/dev_101_services#entity-services